api(v3): add ISO8601 start_time/end_time to POI timeSlots (keep legacy start/end) #3943

Magz94 opened the pull request:
### Short description

Expose ISO-8601 offset-aware times for POI opening hours in API v3 behind an opt-in flag so legacy responses stay unchanged.

### Proposed changes
- **Add (opt-in)** `start_time` / `end_time` to each `timeSlot` as ISO-8601 strings with numeric offset (e.g., `09:00:00+02:00`), derived from the region’s timezone.  
- **Keep** legacy `start` / `end` fields as local wall time (`HH:MM` or `HH:MM:SS`) to avoid breaking existing clients.  
- **Add (opt-in)** `timezone` (IANA key, e.g., `Europe/Berlin`) to each `timeSlot` for clarity/debugging and future client logic.
- New fields are returned only when the query param  `?iso_times=1`  is provided.

Example slot:
```json
{
  "start": "09:00",
  "end": "12:30",
  "start_time": "09:00:00+02:00",
  "end_time": "12:30:00+02:00",
  "timezone": "Europe/Berlin"
}
```

### Side effects
- None expected. This change is **additive** only (no removals/renames), so current consumers using `start`/`end` continue to work unchanged.
- No database migrations. No changes to permissions or caching behavior.

### Faithfulness to issue description and design
Provides offset-aware times per slot (plus the region’s IANA timezone) so clients can implement “open now” logic across time zones/DST, while preserving backward compatibility by default.

### How to test
**Prereq:** run the CMS locally and ensure the `augsburg` region (or any region with POIs and opening hours) is available.

1. Default (legacy only — no ISO fields):
```bash
curl -sS "http://localhost:8000/api/v3/augsburg/de/locations/" \
| python -c 'import sys,json
d=json.load(sys.stdin)
for p in d:
    oh=p.get("opening_hours") or []
    slots=[s for day in oh for s in (day.get("timeSlots") or [])]
    if slots:
        print(p["title"], "—", p["path"])
        for s in slots[:4]:
            print(" ", s.get("start"), "→", s.get("end"),
                  "|", s.get("start_time"), "→", s.get("end_time"),
                  "|", s.get("timezone"))
        break'
```

2. Opt-in ISO fields:
```bash
curl -sS "http://localhost:8000/api/v3/augsburg/de/locations/?iso_times=1" \
| python -c 'import sys,json
d=json.load(sys.stdin)
for p in d:
    oh=p.get("opening_hours") or []
    slots=[s for day in oh for s in (day.get("timeSlots") or [])]
    if slots:
        print(p["title"], "—", p["path"])
        for s in slots[:4]:
            print(" ", s.get("start"), "→", s.get("end"),
                  "|", s.get("start_time"), "→", s.get("end_time"),
                  "|", s.get("timezone"))
        break'
```

**Expected:**  
When calling **without** the flag, each `timeSlot` shows only legacy wall times:
- `start` / `end` (wall time `HH:MM` or `HH:MM:SS`)

When calling **with**`?iso_times=1`, each `timeSlot` additionally includes:
- `start_time` / `end_time` (ISO-8601 with numeric offset)
- `timezone` (IANA key, e.g., `Europe/Berlin`)

**Validation:**
- The region’s configured timezone matches the `"timezone"` field in the response.
- `start_time` / `end_time` use the correct **current** UTC offset for that timezone (incl. DST).
- Legacy wall times `start` / `end` remain unchanged.



<details>
<summary><strong>Screenshots for Validation when calling with the iso Flag </strong> (click to expand)</summary>

**1. Situation**  
<img width="498" height="145" alt="1" src="https://github.com/user-attachments/assets/6a27e288-c9b6-40db-87e4-c6b4d410e02e" />
<img width="628" height="130" alt="1t" src="https://github.com/user-attachments/assets/8ddfd146-270a-45a1-baf6-7d8a265ef2c5" />

**2. Situation**  
<img width="503" height="146" alt="2" src="https://github.com/user-attachments/assets/039549e9-9488-4e2d-8f5f-4fb8f3a389a8" />
<img width="591" height="129" alt="2t" src="https://github.com/user-attachments/assets/44071f18-aae8-46c8-9f35-68a8d09f5a18" />

**3. Situation**  
<img width="495" height="152" alt="3" src="https://github.com/user-attachments/assets/6b5e86be-6292-4101-8558-646c9c7d649e" />
<img width="569" height="130" alt="3t" src="https://github.com/user-attachments/assets/04d50ad7-174b-4d21-9dbf-385d88dc9e83" />

</details>





### Resolved issues
Fixes: #3943

__________________________________________________
[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

Magz94 commented:
Hello everyone ,
I fixed the issues that came such as Ruff,Prettier etc.. but  after adding three new fields to timeSlots (start_time, end_time, timezone in ISO-8601), the API tests fail because fixtures don’t include them.
Proposed solution: keep the default response unchanged and make these fields opt-in via 'iso_times=1'
Is this approach okay, or would you prefer a different solution 
I would be happy to hear your feedback and tell me what to do next, in order to get this merged!

PeterNerlich commented on integreat_cms/api/v3/locations.py:
```diff
@@ -103,7 +161,32 @@ def transform_poi_translation(poi_translation: POITranslation) -> dict[str, Any]
+            new_day["timeSlots"] = new_slots
+            enriched_days.append(new_day)
+
+        opening_hours = enriched_days
```
If we don't want to include the iso time field, we don't need to go through these loops at all, so I would like to pull the condition out.

Additionally, we prefer [comprehensions](https://docs.python.org/3/howto/functional.html#generator-expressions-and-list-comprehensions) wherever they can be used, often times it looks cleaner than doing it imperatively. Combined with non-destructively merging dictionaries with `|` we can simplify this to a surprising degree:
```suggestion
        if not include_iso_times:
            opening_hours = poi.opening_hours
        else:
            # Enrich timeSlots with ISO-8601 times using the region's timezone (opt-in)
            default_tz: str = getattr(settings, "TIME_ZONE", "Europe/Berlin")
            tz_key: str = _tz_key(region_tz_name) or default_tz

            opening_hours = [
                day | {
                    "timeSlots": [
                        slot | {
                            "start_time": _iso_local_time(slot.get("start"), tz_key),
                            "end_time": _iso_local_time(slot.end("end"), tz_key),
                            "timezone": tz_key,
                        }
                        for slot in day.get("timeSlots", [])
                    ]
                }
                for day in poi.opening_hours or []
            ]
```
This form might be a bit confusing for anyone not already familiar with it, but you got to agree that it looks simpler and cleaner. It uses less than 45% as much non-whitespace characters as the original *(yes I know we're not playing code golf, but I just had to point this out :grin:)*

PeterNerlich replied:
Aah, I get it. Yes, in that case we would say the intention of the code changed, so this warrants changing the tests to reflect that new intention.

Magz94 commented on integreat_cms/api/v3/locations.py:
```diff
@@ -188,9 +271,28 @@ def locations(
-        if translation := poi.get_public_translation(language_slug):
-            result.append(transform_poi_translation(translation))
+        translation = poi.get_public_translation(language_slug)
+        if translation:
```
I'm not sure why I took walrus operator out, I think that I was just getting headaches because of those tests, and despaired on trying everything 😆 
I'm taking it back, thank you once again 😄

Magz94 commented:
> Now, I'm sure I've pretty much bulldozed you with that huge wall of text 😅 Please do feel free to enter discussion where you can't follow why a certain thing is supposedly better or worse – even if that is a part that you decide to scrap anyways in order to implement a different approach. That's how we learn 😁 I'll do my best to stay responsive so you don't end up tinkering on this PR forever because of all the down time waiting for responses

@PeterNerlich 
Thanks for the detailed review! I’ve updated the PR to align with your points:
What changed:
* No GET flag: removed `iso_times`, we now always add the extra fields while keeping legacy fields intact.
* Backward compatibility: `start`/`end` remain unchanged; new fields are added: `start_time`, `end_time`, `timezone`.
* Validated timezone: introduced `_ensure_zoneinfo()` and now pass around a real ZoneInfo instance (not a string).
* DST-aware semantics: `start_time` and `end_time` are computed for the next occurrence of each weekday in the region TZ so the numeric offset reflects DST transitions (including “today” if it’s the matching weekday).
* Docs updated: `docs/src/api-docs.rst` now documents the three new fields and includes a note on the “next-occurrence” (DST) semantics.

Style note:
* I originally experimented with a tiny lambda/IIFE inside the list comprehension to keep things super compact 🙈 but Ruff (PLC3002) flagged it. I switched to a comprehension that binds `anchor` via a generator clause, so it’s still clean and passes linting.
This was the lambda version:
```typescript
        opening_hours = [
            (lambda a, d=day: d | {
                "timeSlots": [
                    slot | {
                        "start_time": _iso_local_time(slot.get("start"), region_tz, on_date=a),
                        "end_time": _iso_local_time(slot.get("end"), region_tz, on_date=a),
                        "timezone": tz_key,
                    }
                    for slot in d.get("timeSlots", [])
                ]
            })(
                _next_occurrence_date(region_tz, (idx + WEEK_START) % 7)
            )
            for idx, day in enumerate(poi.opening_hours or [])
        ]
```

I’m happy to tweak anything further. Also: totally open to follow-ups on whether we should eventually add the same ISO fields for contact opening hours as well, but I kept this PR scoped to the location opening hours to avoid surprises.
Thanks again for the thoughtful feedback, really helpful! 😊
